various optimizations, huge changes, WIP

GillesDuvert opened the pull request:
in the course of testing loops speed I found:
1.  that loop speed is not a real problem
2. but that specific algorithms (here is the case for many STRxxx functions such as STRTRIM, but also REBIN, etc...) can be improved
3. but more important, that resorting to openmp in very many places in GDL was not done 100% correctly, inducing noticeable delay in the treatment of small simple operations. 
More specifically, it occured to me that parallelization, using OPENMP and the 'if' clause, was not the best approach in many cases, and was definitely not right in all places.

The if clause was wrong in two ways: 
  1. it would not test if tpool_nthread==1 , where it should just not parallelize at all
  2.  the test on tpool_max_elts is wrong (probably a bad cut and paste that has been propagated everywhere)

Besides, experiences show that it is much faster to avoid entering a #pragma omp parallel section if parallelization is not needed at all.
The main purpose of this PR is institue a new scheme:
 - parallelizing a loop must be done thus:
```
      bool parallelize= yes or no depending on various values, see below
      if (!parallelize) { //which is the most frequent
        for (...) // the loop
      } else { //jump in the same loop, but inside a parallel section:
#pragma omp parallel for num_threads(CpuTPOOL_NTHREADS) if (parallelize) (... eventually other necessary openmp clauses)
        for (...)
      }
```

The "if (parallelize)" above is unnecessary of course but may be handy in case the surrounding 'if (!parallelize) {} else {] ' disappears.

I do not understand why openmp+the if clause is blatantly slower than no openmp at all, but this is the repeated case on my machine, gcc+libgomp 8.4.0 using numerous tests (by adding or removing the #pragma clause etc...)

The minimal test for 'parallelize' is parallelize = (CpuTPOOL_NTHREADS > 1 && nE >= CpuTPOOL_MIN_ELTS && (CpuTPOOL_MAX_ELTS == 0 || CpuTPOOL_MAX_ELTS >= nE));

GillesDuvert committed:
renamed MakeRank to InsureRankAtLeast which is more to the point.

GillesDuvert committed:
wrote a test_rebin that may help to find regressions between successive versions of GDL. Total match of IDL's results is apparently not possible, the roundoff errors cannot be propagated the same as in IDL.

GillesDuvert committed:
noticeable speedup of REBIN.

First attempt to patch inadequate parallelization code spread throughout all GDL:

Found that parallelization, using OPENMP and the 'if' clause, was not the best approach in many cases, and was definitely not right in all places:
- the if clause was wrong in two ways: 
 1) it would not test if tpool_nthread==1 , where it should just not parallelize, bang!
 2) the test on tpool_max_elts is wrong (probably a bad cut and paste that has been propagated everywhere)

Besides, experiences show that it is much faster to avoid entering a #pragma omp parallel section if parallelization is not needed at all.
This commit, and furher, should institue a new scheme:

 - parallelizing a loop must be done thus:
      bool parallelize= yes or no depending on various values, see below
      if (!parallelize) { //which is the most frequent
        for (...) // the loop
      } else { //jump in the same loop, but inside a parallel section:
#pragma omp parallel for num_threads(CpuTPOOL_NTHREADS) if (parallelize) (... eventually other necessary openmp clauses)
        for (...)
      }

The "if (parallelize)" above is unnecessary of course but may be handy in case the surrounding 'if (!parallelize) {} else {] ' disappears.

I do not understand why openmp+the if clause is blatantly slower than no openmp at all, but this is the repeated case on my machine, gcc+libgomp 8.4.0 

The minimal test for 'parallelize' is parallelize = (CpuTPOOL_NTHREADS > 1 && nE >= CpuTPOOL_MIN_ELTS && (CpuTPOOL_MAX_ELTS == 0 || CpuTPOOL_MAX_ELTS >= nE));

GillesDuvert commented:
OK, not unexpected failures. WIP.

GillesDuvert committed:
oops null dimensions must be set to 1 in compression but also expansion.

GillesDuvert committed:
first part of rewriting GDL fundamentals to separate code used when parallelism is needed, (implies some loss of time in the openmp thread code) and the same code without openmp that is well optimised and is faster --- until the benefit of parallelism, that is.

GillesDuvert committed:
revamping parallel sections --- better speed for small numbers

GillesDuvert committed:
finsihed putting all the parallelizable code in the same shape, described in README.THREADPOOL

GillesDuvert committed:
Introduced new types of functions & procedures to differentiate those using the ThreadPool (so, having minimally the 3 thread pool keywords TPOOL_MAX_ELTS, TPOOL_MIN_ELTS, and TPOOL_NOTHREAD ) with others. The 3 keywords are now automatically added to the WarnKey list, since they are not actually taken care of, just 'accepted'. For DLibFunDirect functions, there will be no warning, as these function happily forget about any keyword at all.
I doubt we'll have to go further and effectively support the 3 TP keywords in the future but here are the bases.

alaingdl commented:
How to help you to *test* that ?

Does it means we will have to make some change in the bench ?

GillesDuvert commented:
@alaingdl well, unless I've accidentaly mistyped something, there should be no difference, except in speed for all functions using !CPU.TPOOL  things, that should be faster for all cases where 
`n_elements is < !CPU.TPOOL_MIN_ELTS`.
My results for 'time_test4' show a GDL on average faster than IDL.
I would suggest to run extensive GDL code in your possession (Planck?), seems to me the best way to cover all my changes. I did that on my side, no probs. (even compilation seems faster, ATV is certainly quicker).

GillesDuvert committed:
adjusted some uncorrect values for min and max number of parameters.

GillesDuvert committed:
added a general speedtes that also may cover some yet uncovered code

GillesDuvert committed:
remove message [X|Y]MARGIN to large (adjusted)

GillesDuvert committed:
parallelized yet a few basic functions that were not (and should be)
started editing code for identical writing of parallel loops: preceded by __TRACEOMP__, only 1 #pragma line unless absolutely necessary, etc.
strangely, corrected a few bugs undetected by compilers.

GillesDuvert committed:
regrouped the parallelize conditions in a common function, with options to facilitate the different treatment between a 'threadPool function' that must obey IDl's rules, and the other functions that GDL decided to parallelize to improve speed (primarily array creation, but also some functions that are not threaded in the iDL version)

GillesDuvert committed:
better way to put all threadpool things

GillesDuvert committed:
normally this finishes all thread pool related changes.
test_all_basic_functions reports the perfomances of nearly all basic and threaded functions/procedures.

GillesDuvert commented:
oops, pb in strjoin

GillesDuvert committed:
corrected 2 typing errors caused by automatisation of the numerous changes in so many lines of code.
Also added functions to test_all_basic_functions.

GillesDuvert committed:
yet another typing error

GillesDuvert committed:
for simple cases, use 'print' even if for real doubles and complex the precision is a bit different wrt IDL. The

GillesDuvert committed:
boosted total_over_dim and product_over_dim (over one dimension). As in many places in GDL, the loops on multidimensional adressing are slow, mostly because of too many intermediate indexes are computed, or the loop index does not walk linearly in the larger array, etc.

GillesDuvert committed:
boosted reverse, probably perfectible.

GillesDuvert committed:
waiting for a change in ToStream() that could permit to change the precision of the printed values, prefered to print the arrays with the correct layout (not one value per line) even if it means for floats, doubles and complexes a lesser number of digits wrt IDL's 'implied print' format

GillesDuvert committed:
avoid transposing when not necessary

GillesDuvert committed:
use eigen for transposition, when this is possible it is faster. Not always the case for other tensor utilities of Eigen::

GillesDuvert committed:
removed code that would get a gain of precision wrt IDL in convol(). This was contradictory twith the necessity to stick to IDL's results. besides, it would lengthen the process.

GillesDuvert committed:
pacify clang compiler (and is indeed better like this)

GillesDuvert committed:
complete previous commit : pacify clang compiler (and is indeed better like this)

GillesDuvert committed:
test if changing type of dimension will please clang compiler.

GillesDuvert committed:
typos, unseen, thanks to clang.

GillesDuvert committed:
typos, unseen, thanks to clang.

GillesDuvert committed:
optimized total() for speed if dimension over which to total is 1 (which happens).

GillesDuvert committed:
added special test for major function that work on a particular dimension.

GillesDuvert commented:
I want to merge this PR, as it passes all tests, mine also.
Speed test show a large improvement, GDL is now consitently faster than IDL on all functions described in [README.THREADPOOL](https://github.com/GillesDuvert/gdl/blob/speedups_including_rebin/README.THREADPOOL)
